File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool Dfs(TreeNode* root, long long min, long long max) {
        if (root == NULL) {
            return true;
        }
        if (min < root->val && root->val < max) {
            return Dfs(root->left, min, root->val) && Dfs(root->right, root->val, max);
        }
        return false;
    }
    bool isValidBST(TreeNode* root) {
        return Dfs(root, 1ll * INT_MIN - 1, 1ll * INT_MAX + 1);
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(root, minValue, maxValue):
            if not root:
                return True
            left = dfs(root.left, minValue, root.val)
            right = dfs(root.right, root.val, maxValue)
            res = left and right and minValue < root.val < maxValue
            return res
        return dfs(root, float('-inf'), float('inf'))